tools/yul_phaser/Program.h

```diff
@@ -0,0 +1,95 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include <tools/yul_phaser/Exceptions.h>
+
+#include <liblangutil/CharStream.h>
+
+#include <libyul/optimiser/OptimiserStep.h>
+#include <libyul/optimiser/NameDispenser.h>
+
+#include <optional>
+#include <set>
+#include <string>
+#include <vector>
+
+namespace solidity::yul
+{
+
+struct AsmAnalysisInfo;
+struct Block;
+struct Dialect;
+class YulString;
+
+}
+
+namespace solidity::phaser
+{
+
+class Program
+{
+public:
+	static Program load(std::string const& _sourcePath);
+	void optimize(std::vector<std::string> const& _optimizationSteps);
+
+	size_t codeSize() const;
+
+private:
+	Program(
+		yul::Dialect const& _dialect,
+		std::shared_ptr<yul::Block> _ast
+	):
+		m_ast(_ast),
+		m_nameDispenser(_dialect, *_ast, s_externallyUsedIdentifiers),
+		m_optimiserStepContext{_dialect, m_nameDispenser, s_externallyUsedIdentifiers}
+	{}
+
+	static langutil::CharStream loadSource(std::string const& _sourcePath);
+	static std::shared_ptr<yul::Block> parseSource(
+		yul::Dialect const& _dialect,
+		langutil::CharStream _source
+	);
+	static std::shared_ptr<yul::AsmAnalysisInfo> analyzeAST(
+		yul::Dialect const& _dialect,
+		yul::Block const& _ast
+	);
+	static yul::Block disambiguateAST(
+		yul::Dialect const& _dialect,
+		yul::Block const& _ast,
+		yul::AsmAnalysisInfo const& _analysisInfo
+	);
+	static void applyOptimizationSteps(
+		yul::OptimiserStepContext& _context,
+		yul::Block& _ast,
+		std::vector<std::string> const& _optimizationSteps
+	);
+	static size_t computeCodeSize(yul::Block const& _ast);
+
+	std::shared_ptr<yul::Block> m_ast;
+	yul::NameDispenser m_nameDispenser;
+	yul::OptimiserStepContext m_optimiserStepContext;
```

Comment:
I think the source of the bug is here: Program is copyable but m_optimiserStepContext contains a reference to m_nameDispenser.

Reply:
Thanks for spotting this. Yeah, the way this class handled copy construction was broken. I fixed this by recreating the context in the copy constructor (instead of making a copy) but this still leaves me with shared `m_ast`. I should clone it but I don't yet know how to do it. I could just walk the tree and do it manually but I bet someone has already implemented it somewhere. For now I just made the class `noncopyable` and I'll get back to this problem later.

This was not the cause of `VarNameCleaner` crashes though. I spent a while debugging it and reading through `SSATransform` which fails on `YulAssertion`. It fails because `NameDispenser` gives it names already present in the code. And that's `VarNameCleaner`'s fault. This step changes names in the code but does not mark them as used in the dispenser.

This is completely reproducible in `yulopti` too. Just run
``` bash
build/test/tools/yulopti test/libyul/yulOptimizerTests/ssaTransform/for_reassign_init.yul
```
and type `hgoalaVaalVVVa`. It will crash.

I guess it was not noticed (or maybe just not considered a problem) before because AFAIK it's just final cleanup that does not affect the structure of the AST (just makes the names cleaner) so it's not likely to be anything but the last step.

How should it be fixed?
1) Make `VarNameCleaner` add newly created names to the dispenser. This leaves the removed names in. Not sure if this is a problem or not.
2) Make `VarNameCleaner` both add and remove names to and from the dispenser. This requires a change in dispenser's API which does not have any methods for removing stuff.
3) Just make sure never to run any other optimisation steps after `VarNameCleaner`. I already made it a special case in the `Chromosome`. But that's risky.

Reply:
The VarNameCleaner should not be considered by this tool - it is not a real optimization step, just as the disambiguator is not. Actually, it does not even guarantee unique names any more, so you would have to run the disambiguator after it.

Reply:
Done and moved to a separate PR: #8214.